### ISSUE TITLE
`if` `case` allows to use `void` type

### ISSUE THREAD

**Issue by Zekfad** [dart-lang/sdk#52408]
Consider following code:
```dart
  void _a;

  // Error: This expression has type 'void' and can't be used.
  if (_a is Null) {} // same goes for switch

  // but here are no error
  if (_a case final Null assigned) {
    print('_a is assigned to Null: $assigned');
  }
```

**Comment by lrhn** [comment]
This seems to have regressed since 3.0.0 for the front-end (according to dartpad.dev stable vs. master).
The analyzer does not give a warning in either case.

Having `void`-typed expressions is allowed in switch case expressions (but then requires the switch to be in a `void`-allowing context).
It should probably not be allowed to have `void` as the matched value type anywhere where the value is actually accessed.
That means it can be allowed for non-binding non-checking patterns, `_`, `var _`, `final _`. And possibly with typed patterns that don't actually check anything, `void _`, `dynamic _`, `Object? _` (but you never *need* those, so why allow it).

(Q.v. https://github.com/dart-lang/language/issues/2907).

**Comment by incendial** [comment]
Should `Never` also be disallowed in the pattern matching context?

**Comment by eernstg** [comment]
> Shouldn't `Never` also be disallowed in the pattern matching context?

If the matched value type is `Never` then it's guaranteed that the pattern matching step will never take place. This is usually handled by reporting dead code. Seems that this doesn't occur in this particular situation, cf. https://github.com/dart-lang/sdk/issues/52409.

**Comment by lrhn** [comment]
Agree, there is never any reason to disallow a `Never` type, but it's usually a good idea to warn about dead code, no matter why it's dead. A `(throw "nope") + 1` is not wrong (the semantics are clear), but it is questionable.
Using a "value" with type `Never` is impossible, but harmless.

Using the value of a `void` expression is the opposite: Possible, but potentially dangerous. It's the hand-grenade of values, because it can be anything, and there is no promise it won't change in the future, because you are not expected to actually look.

The only good use of a `void`-typed value is to ignore it. That's why the pattern `_` should be allowed, in case you end up with `(int, void, String) x = ...; switch (x) { case (var i, _, var s): print("$i:$s");`, or `Map<String, void> m = ...; switch (m) { case {"foo": _, "bar": _}: print("Has foo and bar keys"); }`.

It should *very rarely* be necessary to have a pattern with `void` as matched value type, because types like the examples here are silly to begin with, but sometimes destructuring something containing a `void` value requires having a pattern for the `void` value. (But if it was `case Foo(voidGetter: _)`, you'd just not add the getter, and if it was a a `List<void>`, ... why? And don't!)

So, my suggestion (@munificent) is:
**If the matched value type of a pattern is `void`, and the pattern is not an identifier pattern whose name is `_`, it should be a "use of `void` value" error.**

The "identifier pattern whose name is `_`" is the same phrasing we use for list element patterns which do not cause a read of the element.

We _could_ be more permissive, and allow some "uses" of a `void` value that we allow for expressions, but most of those uses were allowed for backwards compatibility reasons, not because it was a good idea:
* Assign `void` to `dynamic` or `void` typed variables, so allowing patterns `dynamic x`, `dynamic(...)` or `void x`.
* Allow casting to anything, so allowing patterns `anyPattern as T` for any pattern and type, "casting away" the `void`-ness.

I'd rather remove that functionality from expressions than repeat the mistake in patterns.

If someone really, really needs to channel a real value through a `void` typed expression, they might just know what they're doing, but the odds are against it. We don't need to make it *easy*, so people can do it by accident.


"If you gaze into the void, your program is fragile and prone to failure."

**Comment by Zekfad** [comment]
Also to mention, there are some weird things with `FutureOr<void>`.

**Comment by munificent** [comment]
> So, my suggestion (@munificent) is:
> **If the matched value type of a pattern is `void`, and the pattern is not an identifier pattern whose name is `_`, it should be a "use of `void` value" error.**

Sounds good to me. :)

**Comment by johnniwinther** [comment]
cc @chloestefantsova

**Comment by lrhn** [comment]
This seems to be quite far-reaching. In the following, all occurrences of `v` after its declaration should be errors. (It's expression positions that are not on the allow-list for having type `void`.)

There are also some cases that are not technically covered by the spec, but which *should* proabably be errors, where a matched value type is `void` and members are invoked on it. (They are extension members, which could be argued to not really be "on" the value, but we don't distinguish extensions `on void` from extensions `on Object?`, so who knows what they do with `this`.

```dart
const void v = "a";

void main() {
  // Invalid uses of `v`.
  if (v case "b") {}
  [if (v case "b") 0];
  switch ("b") {
    case v:
    case == v:
    case != v:
    case < v:
    case <= v:
    case > v:
    case >= v:
    case (v, v): // Constant value in composite pattern.
    case [v]:
    case {v: v}:
    case Object(runtimeType: v):
  }
  (switch ("b") {
    v => 0,
    == v => 0,
    != v => 0,
    < v => 0,
    <= v => 0,
    > v => 0,
    >= v => 0,
    (v, v) => 0,
    [v] => 0,
    {v: v} => 0,
    Object(runtimeType: v) => 0,
    _ => 1,
  });

  // Invocations on values of type `void`.
  // Matched value type that `name` is accessed on is `void`.
  switch ("a") {
    case Void(name: "b"):
    case Object(asVoid: 1):
    case Object(asVoid: []):
    case Object(asVoid: {}):
    case Object(asVoid: == 1):
    case Object(asVoid: != 1):
    case Object(asVoid: < 1):
    case Object(asVoid: <= 1):
    case Object(asVoid: > 1):
    case Object(asVoid: >= 1):  // Any pattern on a `void` receiver, other than `_` and `void x`.
  }
  if ("a" case Void(name: "b")) {}
}

typedef Void = void; // Can't use `void(...)` as a pattern.

extension on void {
  bool operator <(Object? _) => false;
  bool operator <=(Object? _) => false;
  bool operator >(Object? _) => false;
  bool operator >=(Object? _) => false;
  String get name => "void";
  void get asVoid => null;
}
```
